src/shared/sleep-config.c

```diff
@@ -108,77 +111,192 @@ int parse_sleep_config(SleepConfig **ret_sleep_config) {
+        if (r < 0)
+                return r;
+
+        r = sd_device_enumerator_add_match_subsystem(e, "power_supply", true);
+        if (r < 0)
+                return r;
 
-        return r <= 5;
+        r = sd_device_enumerator_add_match_property(e, "POWER_SUPPLY_TYPE", "Battery");
+        /* No battery found check */
+        if (r < 0)
+                return log_error_errno(r, "Failed to get property match to battery: %m");
+
+        *ret = TAKE_PTR(e);
+
+        return 0;
 }
 
 /* Battery percentage capacity fetched from capacity file and if in range 0-100 then returned */
-int read_battery_capacity_percentage(void) {
-        _cleanup_free_ char *bat_cap = NULL;
+static int read_battery_capacity_percentage(sd_device *dev) {
+        const char *power_supply_capacity;
         int battery_capacity, r;
 
-        r = read_one_line_file("/sys/class/power_supply/BAT0/capacity", &bat_cap);
+        assert(dev);
+
+        r = sd_device_get_property_value(dev, "POWER_SUPPLY_CAPACITY", &power_supply_capacity);
         if (r == -ENOENT)
-               return log_debug_errno(r, "/sys/class/power_supply/BAT0/capacity is unavailable. Assuming no battery exists: %m");
+                return log_device_debug_errno(dev, r, "battery capacity is unavailable. Assuming no battery exists: %m");
         if (r < 0)
-               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/capacity: %m");
+                return log_device_debug_errno(dev, r, "Failed to read battery capacity: %m");
 
-        r = safe_atoi(bat_cap, &battery_capacity);
+        r = safe_atoi(power_supply_capacity, &battery_capacity);
         if (r < 0)
-               return log_debug_errno(r, "Failed to parse battery capacity: %m");
+                return log_debug_errno(r, "Failed to parse battery capacity: %m");
 
         if (battery_capacity < 0 || battery_capacity > 100)
-               return log_debug_errno(SYNTHETIC_ERRNO(ERANGE), "Invalid battery capacity");
+                return log_debug_errno(SYNTHETIC_ERRNO(ERANGE), "Invalid battery capacity");
 
         return battery_capacity;
 }
 
+/* If battery percentage capacity is less than equal to 5% return success */
+int battery_is_low(void) {
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        sd_device *dev;
+        bool c = false;
+        int r;
+
+         /* We have not used battery capacity_level since value is set to full
+         * or Normal in case acpi is not working properly. In case of no battery
+         * 0 will be returned and system will be suspended for 1st cycle then hibernated */
+
+        r = get_sd_device_battery(&e);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get available battery: %m");
+
+        FOREACH_DEVICE(e, dev) {
+                r = read_battery_capacity_percentage(dev);
+                if (r < 0) {
+                        log_debug_errno(r, "Failed to get battery capacity: %m");
```

Comment:
`log_device_debug_errno()` is better. Also, please add `", ignoring"` to the log message.

Reply:
`read_battery_capacity_percentage()` will take care of that. since the logging here may be related to -

- failure to get battery current capacity
- failure to convert capacity from string to int
- failure when the battery capacity is in invalid range
so i have left it as debug only.
Hope i am thinking right. You agree ?

Reply:
I suggest `log_device_debug_errno()` instead of `log_debug_errno()`. I do not say anything about the level of logging, and debug level is OK.

Reply:
fixed